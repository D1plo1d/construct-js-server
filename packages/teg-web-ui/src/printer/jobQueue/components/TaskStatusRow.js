import React, { useState, useEffect, useCallback } from 'react'

import IconButton from '@material-ui/core/IconButton'
import Typography from '@material-ui/core/Typography'
import LinearProgress from '@material-ui/core/LinearProgress'
import CircularProgress from '@material-ui/core/CircularProgress'
// import Box from '@material-ui/core/Box'
import Fade from '@material-ui/core/Fade'
import Cancel from '@material-ui/icons/Cancel'
import Play from '@material-ui/icons/PlayArrow'
import Pause from '@material-ui/icons/Pause'

import useConfirm from '../../../common/_hooks/useConfirm'

// const taskColor = (status) => {
//   switch (status) {
//     case 'ERRORED':
//     case 'CANCELLED': {
//       return 'error'
//     }
//     default: {
//       return 'index'
//     }
//   }
// }

const TaskStatusRow = ({
  task,
  cancelTask,
  machineStatus,
  pausePrint,
  resumePrint,
}) => {
  const confirm = useConfirm()

  const confirmedCancelTask = confirm(() => ({
    fn: async () => {
      await cancelTask({
        variables: { machineID: task.machine.id.replace('rust-', '') },
      })
    },
    title: 'Are you sure you want to cancel this print?',
    description: 'You will not be able to resume this print once it is cancelled.',
  }))

  const togglePause = useCallback(() => {
    const mutationFn = task.paused ? resumePrint : pausePrint
    mutationFn({
      variables: {
        taskID: task.id,
      },
    })
  })

  const [, setEtaUpdateCounter] = useState(0)
  useEffect(() => {
    const interval = setInterval(() => {
      setEtaUpdateCounter(i => i + 1)
    }, 15 * 1000)
    return () => clearInterval(interval)
  }, [])

  const isPrinting = ['SPOOLED', 'STARTED'].includes(task.status)
  const blockingHeater = task.machine.components
    // .find(c => c.heater)
    .find(c => c.heater?.blocking)
  // console.log({blockingHeater}, task.machine.components)

  let etaStr = ''

  if (task.startedAt != null && task.estimatedPrintTimeMillis != null) {
    const etaMS = Date.parse(task.startedAt) + task.estimatedPrintTimeMillis - Date.now()
    const eta = new Date(Math.abs(etaMS))
    const hours = Math.floor(Math.abs(etaMS) / 1000 / 3600)
    // const hoursStr = hours > 0 ? `${hours}:` : ''
    const hoursStr = hours > 0 ? `${hours}hr${hours > 1 ? 's ' : ' '}` : ''

    // const twoDigits = n => `${n >= 10 ? '' : '0'}${n}`

    etaStr = (() => {
      // const timeStr = `${hoursStr}${twoDigits(eta.getMinutes())}:${twoDigits(eta.getSeconds())}`
      const timeStr = `${hoursStr}${eta.getMinutes()}m`
      if (etaMS < 0) {
        return ` (${timeStr} over time estimate)`
      }
      return ` (${timeStr} remaining)`
    })()
  }

  const disabled = !['READY', 'PRINTING', 'PAUSED'].includes(machineStatus)

  let heaterOverlay = <div/>
  if (isPrinting && blockingHeater != null) {
    const { actualTemperature, targetTemperature } = blockingHeater.heater

    const temperaturePercent = Math.min((
      100 * actualTemperature / targetTemperature
    ) || 0, 100)

    heaterOverlay = (
      <div
        style={{
          display: 'flex',
          // flexDirection: 'column',
          justifyContent: 'center',
          alignItems: 'center',
          // position: 'relative',
          // marginTop: 'calc(-1rem - 16px)',
          marginTop: -16,
          background: 'white',
          // background: 'rgba(255, 255, 255, .8)',
          zIndex: 1,
        }}
      >
        <CircularProgress
          variant="determinate"
          value={temperaturePercent}
          size={16*3}
        />
        {/* <Box
          top={0}
          left={0}
          bottom={0}
          right={0}
          position="absolute"
          display="flex"
          alignItems="center"
          justifyContent="center"
        >
        </Box> */}
        <Typography
          variant="h6"
          component="div"
          color="textSecondary"
          style={{
            // marginTop: 8,
            marginLeft: 16,
          }}
        >
          {`Heating ${blockingHeater.name} `}
          {/* {`(${Math.round(actualTemperature)} / ${Math.round(targetTemperature*10)/10}°C)`} */}
          {`to ${Math.round(targetTemperature*10)/10}°C`}
        </Typography>
      </div>
    )
  }

  return (
    <div
      style={{
        display: 'grid',
        gridTemplateRows: '1fr',
        gridTemplateAreas: '"a"',
      }}
    >
      <Fade
        style={{
          transitionDelay: `0ms`,
          gridArea: 'a',
        }}
        timeout={ 800 }
        in={isPrinting && blockingHeater != null}
      >
        { heaterOverlay }
      </Fade>
      <div style={{
        gridArea: 'a',
      }}>
        <Typography
          variant="body2"
          gutterBottom
          style={{
            marginBottom: 0,
          }}
        >
          {
            (() => {
              const taskOnMachine = `on ${task.machine.name}`
              // console.log(task.status)

              if (['CANCELLED', 'ERRORED'].includes(task.status)) {
                const stopptedAtText = new Date(Date.parse(task.stoppedAt)).toLocaleString()
                const statusWord = (
                  task.status === 'CANCELLED' ? 'Cancelled' : 'Errored'
                )
                return `${statusWord} ${taskOnMachine} at ${stopptedAtText}`
              }
              if (['FINISHED', 'PAUSED'].includes(task.status)) {
                const stopptedAtText = new Date(Date.parse(task.stoppedAt)).toLocaleString()
                return `Print ${task.status.toLowerCase()} ${taskOnMachine} at ${stopptedAtText}`
              }
              return `Printing ${taskOnMachine}`
            })()
          }
        </Typography>

        <div
          style={{
            display: 'flex',
            alignItems: 'center',
          }}
        >
          { !task.settled && (
            <Typography
              variant="body2"
              style={{
                marginRight: 12,
              }}
            >
              {task.percentComplete.toFixed(1)}
              %
              {etaStr}
            </Typography>
          )}
          <div
            style={{
              flexGrow: 1,
            }}
          >
            <LinearProgress
              variant="determinate"
              value={task.percentComplete}
            />
          </div>
          { !task.settled && (
            <>
              <IconButton
                aria-label={task.paused ? 'resume print' : 'pause print'}
                disabled={
                  ['CANCELLED', 'ERROR', 'DONE'].includes(task.status) || disabled
                }
                onClick={togglePause}
                style={{
                  marginTop: -12,
                  marginBottom: -12,
                }}
              >
                {task.paused ? <Play /> : <Pause />}
              </IconButton>
              <IconButton
                aria-label="cancel"
                disabled={
                  ['CANCELLED', 'ERROR', 'DONE'].includes(task.status) || disabled
                }
                onClick={confirmedCancelTask}
                style={{
                  marginTop: -12,
                  marginBottom: -12,
                  marginRight: -14,
                  textAlign: 'right',
                }}
              >
                <Cancel />
              </IconButton>
            </>
          )}
        </div>
      </div>
    </div>
  )
}

export default TaskStatusRow
