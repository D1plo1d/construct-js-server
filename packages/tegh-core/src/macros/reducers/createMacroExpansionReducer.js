import { loop, Cmd } from 'redux-loop'
import { Record } from 'immutable'

import isMacroEnabled from '../../config/selectors/isMacroEnabled'
import { PREEMPTIVE } from '../../spool/types/PriorityEnum'

import { SET_CONFIG } from '../../config/actions/setConfig'
import { DESPOOL_TASK } from '../../spool/actions/despoolTask'

import spoolTask from '../../spool/actions/spoolTask'
import requestDespool from '../../spool/actions/requestDespool'

const initialState = Record({
  config: null,
  enabled: false,
})()

const createMacroExpansionReducer = (
  meta,
  macroFn,
) => (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = state.payload
      const enabled = isMacroEnabled({ config, meta })

      return state.merge({
        config,
        enabled,
      })
    }
    case DESPOOL_TASK: {
      const { macro, args, task } = action.payload

      if (macro === meta.macro && state.enabled) {
        // expand the macro into it's expanded gcode (an array of strings)
        const data = macroFn(args, state)
        /*
         * 1. insert the expanded gcode lines in a task before everything else
         * effectively swapping the host macro for the expanded gcode lines
         * 2. despool the first line of that expanded gcode
         */
        const taskAttrs = {
          name: 'test.ngc',
          internal: task.internal,
          priority: PREEMPTIVE,
          data,
        }
        const actions = [
          Cmd.action(spoolTask(taskAttrs, { prepend: true })),
          Cmd.action(requestDespool()),
        ]

        loop(state, Cmd.list(actions))
      }

      return state
    }
    default: {
      return state
    }
  }
}

export default createMacroExpansionReducer
