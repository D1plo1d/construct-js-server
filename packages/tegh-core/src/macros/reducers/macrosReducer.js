// import { loop, Cmd } from 'redux-loop'
// import { Record, Map } from 'immutable'
//
// import { SET_CONFIG } from '../../config/actions/setConfig'
// import { SPOOL_MACRO } from '../../spool/actions/spoolMacro'
// import spoolTask from '../../spool/actions/spoolTask'
//
// import { NORMAL } from '../../spool/types/PriorityEnum'
//
// import getMacroRunFnsByName from '../../pluginManager/selectors/getMacroRunFnsByName'
//
// export const initialState = Record({
//   config: null,
//   macros: Map(),
// })()
//
// const macrosReducer = (state = initialState, action) => {
//   switch (action.type) {
//     case SET_CONFIG: {
//       const nextState = initialState
//         .set('macros', getMacroRunFnsByName(action.payload))
//         .set('config', action.payload.config)
//
//       return nextState
//     }
//     case SPOOL_MACRO: {
//       const {
//         internal,
//         priority,
//         macro,
//         args,
//       } = action.payload
//
//       const macroRunFn = state.macros.get(macro)
//
//       if (macroRunFn == null) {
//         throw new Error(`Macro ${macro} does not exist`)
//       }
//
//       return loop(state, Cmd.action(
//         spoolTask({
//           name: macro,
//           internal,
//           priority: priority || macroRunFn.priority || NORMAL,
//           data: macroRunFn(args, { config: state.config }),
//         }),
//       ))
//     }
//     default: {
//       return state
//     }
//   }
// }
//
// export default macrosReducer
