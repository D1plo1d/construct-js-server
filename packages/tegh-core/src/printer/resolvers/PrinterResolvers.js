import isIdle from '../../spool/selectors/isIdle'
import getComponents from '../../config/selectors/getComponents'
import getComponentsState from '../selectors/getComponentsState'
import getPluginModels from '../../config/selectors/getPluginModels'
import ComponentTypeEnum from '../../config/types/components/ComponentTypeEnum'

const PrinterResolvers = {
  Printer: {
    id: source => source.config.printer.id,
    name: source => getPluginModels(source.config).getIn(['tegh-core', 'name']),
    targetTemperaturesCountdown: source => getComponentsState(source).targetTemperaturesCountdown,
    activeExtruderID: source => getComponentsState(source).activeExtruderID,
    enabledMacros: source => source.macros.enabledMacros,
    error: source => source.status.error,

    components: (source, args) => {
      const id = args.componentID
      const components = getComponents(source.config)

      if (id != null) {
        const component = components.get(id)
        if (component == null) {
          throw new Error(`Component ID: ${id} does not exist`)
        }
        return [component]
      }
      return components
        .toList()
        .sortBy(c => (
          `${ComponentTypeEnum.indexOf(c.type)}${c.model.get('name')}`
        ))
    },

    plugins: (source, args) => {
      const { plugins } = source.config.printer

      if (args.package != null) {
        const plugin = plugins.find(p => p.package === args.package)
        if (plugin == null) {
          throw new Error(`Plugin package: ${args.package} does not exist`)
        }
        return [plugin]
      }

      return plugins
    },

    status: (source) => {
      if (!isIdle(source.spool)) return 'PRINTING'
      const { status } = source.status
      return status.substring(status.lastIndexOf('/') + 1)
    },

    logEntries: (source, args) => {
      let entries = source.log.get('logEntries')
      if (args.level != null) {
        entries = entries.filter(log => log.level === args.level)
      }
      if (args.source != null) {
        entries = entries.filter(log => log.source === args.source)
      }
      return entries.toArray()
    },
  },
}

export default PrinterResolvers
