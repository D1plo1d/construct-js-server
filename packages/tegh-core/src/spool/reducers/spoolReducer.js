import { loop, Cmd } from 'redux-loop'
import {
  Record, List, Map,
} from 'immutable'

import isIdle from '../selectors/isIdle'
import getEnabledHostMacros from '../../pluginManager/selectors/getEnabledHostMacros'

import {
  EMERGENCY,
  NORMAL,
  PREEMPTIVE,
  priorityOrder,
} from '../types/PriorityEnum'

import {
  PRINTING,
} from '../types/TaskStatusEnum'

/* config actions */
import { SET_CONFIG } from '../../config/actions/setConfig'
/* printer actions */
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'
/* job actions */
import { DELETE_JOB } from '../../jobQueue/actions/deleteJob'
/* task actions */
import { SPOOL_TASK } from '../actions/spoolTask'
import requestDespool, { REQUEST_DESPOOL } from '../actions/requestDespool'
import despoolTask from '../actions/despoolTask'
import { CANCEL_TASK } from '../actions/cancelTask'

export const initialState = Record({
  priorityQueues: Record({
    [EMERGENCY]: List(),
    [PREEMPTIVE]: List(),
    [NORMAL]: List(),
  })(),
  tasks: Map(),
  currentTaskID: null,
  enabledHostMacros: List(),
})()

const removeTaskReferences = (state) => {
  const nextTaskIDs = state.tasks.map(task => task.id)

  let nextState = state
  priorityOrder.forEach((priority) => {
    nextState = nextState.updateIn(['priorityQueues', priority], queue => (
      queue.filter(taskID => nextTaskIDs.includes(taskID))
    ))
  })

  if (nextTaskIDs.includes(state.currentTaskID) === false) {
    nextState = nextState.set('currentTaskID', null)
  }

  return nextState
}

const spoolReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      return initialState
        .set('enabledHostMacros', getEnabledHostMacros(action.payload))
    }
    /* Spool reset actions */
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR: {
      return initialState
        .set('enabledHostMacros', state.enabledHostMacros)
    }
    case DELETE_JOB: {
      const { jobID } = action.payload

      const nextState = state.update('tasks', tasks => (
        tasks.filter(task => task.jobID !== jobID)
      ))

      return removeTaskReferences(nextState)
    }
    case CANCEL_TASK: {
      const { taskID } = action.payload

      const nextState = state.update('tasks', tasks => tasks.remove(taskID))

      return removeTaskReferences(nextState)
    }
    case SPOOL_TASK: {
      const { task, prepend } = action.payload
      const { id, priority } = task

      let nextState = state

      /*
      * if the task is an emergency then cancel all other tasks in the queue
      */
      if (priority === EMERGENCY) {
        nextState = initialState
          .set('enabledHostMacros', state.enabledHostMacros)
      }

      if (isIdle(nextState) === false && task.internal !== true) {
        throw new Error('Cannot spool non-emergency tasks when printing a job')
      }

      /* add the task to the spool */
      nextState = nextState
        .setIn(['tasks', id], task)
        .updateIn(['priorityQueues', priority], (list) => {
          if (prepend) {
            return list.unshift(id)
          }
          return list.push(id)
        })

      /*
       * despool the first line if nothing is spooled
       */
      const shouldDespool = nextState.currentTaskID == null
      if (shouldDespool) {
        return loop(nextState, Cmd.action(requestDespool()))
      }

      return nextState
    }
    case REQUEST_DESPOOL: {
      const { priorityQueues, currentTaskID } = state
      let nextState = state
      const currentTask = nextState.tasks.get(currentTaskID)

      if (currentTask != null) {
        /*
         * despool the next line or finish the task
         */
        if (currentTask.currentLineNumber < currentTask.data.size - 1) {
          /*
           * if the task has more lines to execute then increment the line number
           */
          nextState = nextState
            .updateIn(['tasks', currentTaskID, 'currentLineNumber'], i => i + 1)
        } else {
          /* delete the task upon completion */
          nextState = nextState
            .set('currentTaskID', null)
            .update('tasks', tasks => tasks.remove(currentTaskID))
        }
      }

      if (nextState.currentTaskID == null) {
        /*
         * if the current task is done then despool the next task if there is
         * anything to despool.
         */
        const priority = priorityOrder.find(priorityOption => (
          priorityQueues[priorityOption].size > 0
        ))

        const allQueuesAreEmpty = priority == null
        if (allQueuesAreEmpty) {
          return initialState
            .set('enabledHostMacros', state.enabledHostMacros)
        }

        const nextTaskID = state.priorityQueues[priority].first()

        /*
         * start the task
         */
        nextState = nextState
          .mergeIn(['tasks', nextTaskID], {
            startedAt: new Date().toISOString(),
            status: PRINTING,
            currentLineNumber: 0,
          })
          .set('currentTaskID', nextTaskID)
          .updateIn(['priorityQueues', priority], list => list.shift())
      }

      const nextTask = nextState.tasks.get(nextState.currentTaskID)
      return loop(
        nextState,
        Cmd.action(despoolTask(nextTask, state.enabledHostMacros)),
      )
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
