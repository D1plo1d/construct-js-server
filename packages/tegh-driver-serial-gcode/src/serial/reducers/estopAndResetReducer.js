import { loop, Cmd } from 'redux-loop'

import {
  estop,
  connectPrinter,
  DESPOOL_TASK,
} from 'tegh-core'

export const initialState = null

const estopAndResetReducer = (state = initialState, action) => {
  switch (action.type) {
    case DESPOOL_TASK: {
      const { macro } = action.payload

      if (macro === 'eStop') {
        return loop(state, Cmd.action(estop()))
      }

      if (macro === 'reset') {
        return loop(state, Cmd.action(connectPrinter()))
      }

      return state
    }
    default: {
      return state
    }
  }
}

export default estopAndResetReducer
