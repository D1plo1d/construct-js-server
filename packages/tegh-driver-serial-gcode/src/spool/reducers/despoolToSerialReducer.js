import { loop, Cmd } from 'redux-loop'
import { Record } from 'immutable'

import {
  DESPOOL_TASK,
  getCurrentLine,
  isEmergency,
  requestDespool,
  driverError,
  parseGCode,
} from 'tegh-core'

import { SERIAL_CLOSE } from '../../serial/actions/serialClose'
import { SERIAL_RECEIVE } from '../../serial/actions/serialReceive'
import serialSend, { SERIAL_SEND } from '../../serial/actions/serialSend'

export const REQUEST_DESPOOL_ON_OK = 'REQUEST_DESPOOL_ON_OK'
export const RESEND_ON_OK = 'RESEND_ON_OK'
export const IGNORE_OK = 'IGNORE_OK'

export const initialState = Record({
  onNextOK: REQUEST_DESPOOL_ON_OK,
  currentSerialLineNumber: 1,
  lastTaskSent: null,
})()

/*
 * Intercepts DESPOOL actions and sends the current gcode line to the
 * printer.
 *
 * Intercepts SERIAL_RECEIVE actions with the serialReceiveReducer
 */
const despoolToSerialReducer = (state = initialState, action) => {
  switch (action.type) {
    case DESPOOL_TASK: {
      const { task, isHostMacro } = action.payload

      if (isHostMacro) return state

      const emergency = isEmergency.resultFunc(task)
      const currentLine = getCurrentLine.resultFunc(task)

      /*
       * Send emergency GCodes without line numbers since the printer may be in
       * an unknown state which may include a line number mismatch.
       *
       * M112 Emergency Stops without a line number so that it will
       * be executed by the printer immediately without the opportunity for a
       * line number mismatch to cause an error and potentially prevent the
       * estop.
       */
      const lineNumber = emergency ? false : state.currentSerialLineNumber

      const nextState = state.set('lastTaskSent', task)

      return loop(
        nextState,
        Cmd.action(serialSend({
          ...parseGCode(currentLine),
          lineNumber,
        })),
      )
    }
    case SERIAL_CLOSE: {
      return state.set('currentSerialLineNumber', 1)
    }
    case SERIAL_SEND: {
      const { lineNumber } = action.payload

      if (typeof lineNumber !== 'number') return state

      return state.set('currentSerialLineNumber', lineNumber + 1)
    }
    case SERIAL_RECEIVE: {
      const { payload } = action
      const task = state.lastTaskSent

      switch (payload.type) {
        case 'ok': {
          switch (state.onNextOK) {
            case REQUEST_DESPOOL_ON_OK: {
              return loop(state, Cmd.action(requestDespool()))
            }

            case RESEND_ON_OK: {
              const previousSerialLineNumber = state.currentSerialLineNumber - 1

              const currentLine = getCurrentLine.resultFunc(task)

              const nextState = state
                .set('onNextOK', IGNORE_OK)

              const nextAction = serialSend({
                ...parseGCode(currentLine),
                lineNumber: previousSerialLineNumber,
              })

              return loop(
                nextState,
                Cmd.action(nextAction),
              )
            }

            case IGNORE_OK: {
              return state.set('onNextOK', REQUEST_DESPOOL_ON_OK)
            }

            default: {
              throw new Error(`Invalid onNextOK state: ${state.onNextOK}`)
            }
          }
        }
        case 'resend': {
          const previousSerialLineNumber = state.currentSerialLineNumber - 1

          /*
           * Tegh only sends one line at a time. If a resend is requested for a
           * different line number then this is likely an issue of the printer's
           * firmware.
           */
          if (payload.lineNumber !== previousSerialLineNumber) {
            throw new Error(
              `resend line number ${payload.lineNumber} does not `
              + `match previous line number ${previousSerialLineNumber}`,
            )
          }
          // wait for the ok sent after the resend (see marlinFixture.js)
          return state
            .set('onNextOK', RESEND_ON_OK)
        }
        case 'error': {
          let { message } = action.payload
          if (task != null) {
            message = `${task.name}:${task.currentLineNumber}: ${message}`
          }
          const errorAction = driverError({
            code: 'FIRMWARE_ERROR',
            message,
          })

          return loop(state, Cmd.action(errorAction))
        }
        default: {
          return state
        }
      }
    }
    default: {
      return state
    }
  }
}

export default despoolToSerialReducer
