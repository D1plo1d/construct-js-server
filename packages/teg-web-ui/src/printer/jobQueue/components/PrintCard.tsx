import React, { useState, useCallback } from 'react'
import { Link } from 'react-router-dom'
import truncate from 'truncate'

import Card from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import CardContent from '@material-ui/core/CardContent'
// import IconButton from '@material-ui/core/IconButton'
// import Menu from '@material-ui/core/Menu'
// import MenuItem from '@material-ui/core/MenuItem'
// import ListItemIcon from '@material-ui/core/ListItemIcon'
// import ListItemText from '@material-ui/core/ListItemText'

// import MoreVert from '@material-ui/icons/MoreVert'
// import Delete from '@material-ui/icons/Delete'
// import Reorder from '@material-ui/icons/Reorder'

import TaskStatusRow from './TaskStatusRow'
import { Typography } from '@material-ui/core'
// import useConfirm from '../../../common/_hooks/useConfirm'
// import { Typography } from '@material-ui/core'

const PrintCard = ({
  print,
  cancelTask,
  pausePrint,
  resumePrint,
  // deletePart,
}) => {
  // const confirm = useConfirm()
  // const [menuAnchorEl, setMenuAnchorEl] = useState()

  // const openMenu = useCallback(event => setMenuAnchorEl(event.target), [])
  // const closeMenu = useCallback(() => setMenuAnchorEl(null), [])

  const { task, part } = print
  const machineStatus = task.machine.status

  const shortName = truncate(part.name, 32)
  // console.log({ task } )

  // const confirmedDeletePart = confirm(() => ({
  //   fn: () => {
  //     deletePart({
  //       variables: {
  //         input: {
  //           partID: task.id,
  //         },
  //       },
  //     })
  //     closeMenu()
  //   },
  //   title: 'Are you sure you want to delete this part?',
  //   description: part.name,
  // }))
  let statusSetAt = new Date(Date.parse(task.stoppedAt || task.startedAt))
    .toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })
    .replace(' a.m.', 'am')
    .replace(' p.m.', 'pm')

  const isPrinting = ['SPOOLED', 'STARTED'].includes(task.status)
  const capitalizedStatus = task.status.charAt(0) + task.status.toLowerCase().slice(1)

  let abortInfo = ''
  if (task.settled && task.status != 'FINISHED') {
    abortInfo = ` (at ${task.percentComplete.toFixed(1)}% complete)`
  }

  return (
    <Card>
      <CardHeader
        title={(
          <Link
            to={`./printing/${task.partID}/`}
            style={{ textDecoration: 'none', color: 'inherit' }}
          >
            {isPrinting && 'Printing ' || capitalizedStatus + ' '}
            {shortName}
            {/* {!isPrinting && ` ${task.status.toLowerCase()}`} */}
          </Link>
        )}
        subheader={
          `${task.stoppedAt ? capitalizedStatus : 'Started'}${abortInfo} at ${statusSetAt}`
          + ` on ${task.machine.name}`}
      />

      {/* <Menu
        id="long-menu"
        anchorEl={menuAnchorEl}
        open={menuAnchorEl != null}
        onClose={closeMenu}
      >
        <MenuItem onClick={confirmedDeletePart}>
          <ListItemIcon>
            <Delete />
          </ListItemIcon>
          <ListItemText primary="Delete Part" />
        </MenuItem>
      </Menu> */}

      {!task.settled && (
        <CardContent
          style={{
            paddingTop: 0,
          }}
        >
          <TaskStatusRow
            task={task}
            key={task.id}
            {...{
              cancelTask,
              pausePrint,
              resumePrint,
              machineStatus,
            }}
          />
        </CardContent>
      )}
    </Card>
  )
}

export default PrintCard
