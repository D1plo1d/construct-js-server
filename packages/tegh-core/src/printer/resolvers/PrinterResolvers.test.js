import { List, Map, Record } from 'immutable'
import typeDefs from 'tegh-schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { CONNECTING } from '../types/statusEnum'
import { INFO } from '../../log/types/logLevelEnum'
import { TOOLHEAD, FAN } from '../../config/types/components/ComponentTypeEnum'

import { initialState as spoolInitialState } from '../../spool/reducers/spoolReducer'
import { initialState as macrosInitialState } from '../../macros/reducers/macrosReducer'
import { initialState as logInitialState } from '../../log/reducers/logReducer'
import { initialState as statusInitialState } from '../reducers/statusReducer'
import { MockConfig } from '../../config/types/Config'
import SchemaForm from '../../pluginManager/types/SchemaForm'

import PrinterResolvers from './PrinterResolvers'

describe('PrinterResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()

    const state = Record({
      config: MockConfig(),
      schemaForms: Record({
        plugins: Map({
          'tegh-core': SchemaForm({
            id: 'tegh-core',
          }),
          'tegh-driver-serial-gcode': SchemaForm({}),
          'tegh-macros-default': SchemaForm({}),
        }),
      })(),
      fixedListComponentTypes: List(['TOOLHEAD']),
      spool: spoolInitialState,
      status: statusInitialState.set('status', CONNECTING),
      macros: macrosInitialState.set('enabledMacros', List(['myMacro'])),
      log: logInitialState.setIn(['logEntries', 0], {
        source: 'SERIAL',
        level: INFO,
        message: 'This is an example log',
      }),
      driver: 'tegh-driver-serial-gcode',
      'tegh-driver-serial-gcode': {
        components: Record({
          targetTemperaturesCountdown: 3,
          activeExtruderID: 'heater1ID',
          byID: Map({
            heater1Address: Record({
              id: 'heater1ID',
              type: TOOLHEAD,
            })(),
            fan1Address: Record({
              id: 'fan1ID',
              type: FAN,
            })(),
          }),
        })(),
      },
    })()

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Printer',
      resolvers: PrinterResolvers,
      rootValue: state,
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
