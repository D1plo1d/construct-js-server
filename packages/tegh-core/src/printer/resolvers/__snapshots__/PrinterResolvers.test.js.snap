// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrinterResolvers presents a consistent API 1`] = `
Object {
  "_testPrinter": Object {
    "activeExtruderID": "heater1ID",
    "components": Array [
      Object {
        "id": "bbbserialController",
      },
      Object {
        "id": "x111",
      },
      Object {
        "id": "y111",
      },
      Object {
        "id": "z111",
      },
      Object {
        "id": "heater_1",
      },
      Object {
        "id": "abvwee1",
      },
      Object {
        "id": "bdfbxb",
      },
      Object {
        "id": "ndrgrwef",
      },
      Object {
        "id": "ndrgrwef2",
      },
    ],
    "enabledMacros": Array [
      "myMacro",
    ],
    "error": null,
    "fixedListComponentTypes": Array [
      "TOOLHEAD",
    ],
    "id": "aaxcvxcvcxv-bvb-csdf234231",
    "logEntries": Array [
      Object {},
    ],
    "name": "Left Home Lulzbot",
    "plugins": Array [
      Object {
        "id": "asgbvas23",
      },
      Object {
        "id": "aaabbbccc333",
      },
      Object {
        "id": "aaabbbccc123",
      },
      Object {
        "id": "asdfv31",
      },
    ],
    "status": "CONNECTING",
    "targetTemperaturesCountdown": 3,
  },
}
`;
